### ISSUE TITLE
How to re-use a VeasyForm with different state that has already been mounted?

### ISSUE THREAD

**Issue by JeremyColton** [Albert-Gao/veasy#23]
My use-case is:
1. Page contains a component that contains a `<VeasyForm>`
2. User clicks on "Add User" to open the form which is shown as a ReactStrap `<Modal>`
3. User closes the form
4. User clicks on an existing User from a table of users, I want to open the `<VeasyForm>` compontent again, but this time showing the state of the selected user.

If I add `componentWillReceiveProps(nextProps)` to the component, I can pass the selected User object. 

How do I convert this easily into a new VeasyForm formSchema JSON?
I then need to call again `createInitialState(this.formSchema, myState)` ?

Or is there a better way to update the state of an already mounted `<VeasyForm>` ?

**Comment by Albert-Gao** [comment]
So, let me know if I get it wrong:

1. You use `<VeasyForm>` to create a user entity.
1. The information will be saved to a list.
1. Later on, the user can click some random user from the list we mentioned in step2 to bring up the same `<VeasyForm>`. But this time, it's not for creating, it's for editing.
1. And in terms of editing, the information may be different for the previous one.
1. And you want to validate this new information as well.

`<VeasyForm>` doesn't design for this use case.  But I can see your point. 

**Maybe** you can find some luck from this https://albert-gao.github.io/veasy/#/diy . This will let you use `Veasy` programmatically, but you need to bind all the props for youself. The benefit is, you now don't need to follow the constraint of `<VeasyForm>`, you can bind any field item anywhere as long as they are inside a form.

Give me some time to think about this. Do you have any suggestion?

**Comment by JeremyColton** [comment]
Your use-case is correct. I'll re-use the same mounted `<VeasyForm>` for adding and updating a User.

I actually got it working by the following:
` componentWillReceiveProps(nextProps) {`

        `let selectedRow = nextProps.selectedRow;

        if (selectedRow) {
            // user clicked on table row to update a user
            this.formSchema.email.default = selectedRow.email;
            this.formSchema.first_name.default = selectedRow.first_name;
            this.formSchema.last_name.default = selectedRow.last_name;
            this.formSchema.password.default = selectedRow.password;
            this.myState.title = "Update User";
        } else {
            //is empty so user clicked on 'add user'
            this.formSchema.email.default = '';
            this.formSchema.first_name.default = '';
            this.formSchema.last_name.default = '';
            this.formSchema.password.default = '';
            this.myState.title = "New User";
        }

        this.state = createInitialState(this.formSchema, this.myState);
    }'

**Comment by Albert-Gao** [comment]
Hmm, or, you can extract this `<VeasyForm>` to another component.
Then pass all the data as props, by this, you could reuse this form.

**Comment by JeremyColton** [comment]
How do I pass the input data as props? The value of the `<TextInput>` is provided by the VeasyForm's config...

**Comment by Albert-Gao** [comment]
You can, because the value you get from the props is in the component state as well.

You can pass the input as props, and in the constructor, you use the `createInitialState` to do generate the state, then do the rendering. But it will lead to useless rendering. Let me think some way better.

**Comment by JeremyColton** [comment]
Can there be an API to programmatically  update a field's value which also triggers the onChange handler for the field?

**Comment by Albert-Gao** [comment]
the upcoming beforeValidation rule seems the answer

**Comment by JeremyColton** [comment]
A new API? 
It will work if the `<TextInput>` uses `defaultValue={value}` instead of `value={value}` ? (due to bug https://github.com/Albert-Gao/veasy/issues/18)

**Comment by Albert-Gao** [comment]
Nope, it solves a different problem.

**Comment by JeremyColton** [comment]
beforeValidation fires after the user types a character (ie after the onKeyUp event) in the given field and before any validation rules?

**Comment by Albert-Gao** [comment]
It's a rule. So, it will be executed when the onChange gets triggered.

But it's a special rule, so it will run before any other rules. Which means, beforeValidation

**Comment by JeremyColton** [comment]
Ok, but I can't use it to set field values on an already-mounted form.

I settled on the solution to update the form schema with new values, eg:
My veasy form was mounted with empty values for email. I then want to re-use my form to edit an existing user's email. So I do:
`this.formSchema.email.default = 'a@a.com'`
`const initialState = createInitialState(formSchema);`

This sets the existing user's email value in the form. Seems to work ok...

**Comment by Albert-Gao** [comment]
Glad to solved it. This is quite hacky. And it will recreate the whole state. I am surprised that it's working. The value is bind through value rather than default value. Right?

You can just change the piece of state 
using setState rather than the whole state. Is better :)

**Comment by JeremyColton** [comment]
Bound through defaultValue. It felt hacky!

Use this.setState({email:'a@a.com'}) won't change the form's state... What am I misunderstanding here?

**Comment by Albert-Gao** [comment]
Yes, the path is not right. Inspect the actual state. should be title.value rather than title

**Comment by JeremyColton** [comment]
I am using the new BaseInput component with the TextInput component.

Using your example app, to render an 'update' form with values pre-set in each field, I set a field's value via the 'default' property:
this.formSchema = {
            title: { minLength: 3, default: 'hiiii'},
            description: { minLength: 5 },
            answer: { default: "not care" }
        };

When I inspect the state, the 'title' is as follows:
title: {
   errorText:""
   status:"normal"
   value:"hiiii"
}

The status should be 'ok' since the field is valid, correct?

**Comment by Albert-Gao** [comment]
I think it has coded that way. When the initial state has been created from that `createInitialState()`, it won't trigger the validation for the `default` value.

But seems it should. hmm...
That means we have to validate the `default` value to make sure that the default value follows the rules as well.

Thanks for pointing it out. I will see and fix it.

**Comment by Albert-Gao** [comment]
No, I don't think it should be triggered. It means when the form is initialized, some of the fields will be marked as `ok` while the user hasn't given any input.

The default value should behave more like a suggestion. What do you think? :)

**Comment by Albert-Gao** [comment]
hmm

But sometimes the `default` value does behave like a default value. Where its purpose is to make the user doesn't need to input something. So, in this case, it should be trigger the validation.

Alright, maybe we should make this default supports something like this:

```javascript
this.formSchema = {
       title: { 
           minLength: 3, 
           default: 'hiiii'
       }
};
```

Then it won't trigger the validation when generating the initial state.

But when you setup like this:

```javascript
this.formSchema = {
       title: { 
           minLength: 3, 
           default: ['hiiii', true]
       }
};
```

Then it will, the second parameter implies `triggerValidation = true`

What do you think? Is there any simpler represetation?

**Comment by Albert-Gao** [comment]
Yes, I agree with you. :)
It's fixed. I will publish after I sort out the new files.

**Comment by Albert-Gao** [comment]
Fixed, please update to 1.5.1 :) Thanks for using!

**Comment by JeremyColton** [comment]
Installed and tested - works great! Thanks for your fast fix! :)
